proxy/freedom/freedom.go

```diff
@@ -383,6 +401,82 @@ func (w *PacketWriter) WriteMultiBuffer(mb buf.MultiBuffer) error {
 	return nil
 }
 
+type NoisePacketWriter struct {
+	Writer     PacketWriter
+	firstWrite bool
+}
+
+// MultiBuffer writer with Noise in first packet
+func (w *NoisePacketWriter) WriteMultiBuffer(mb buf.MultiBuffer) error {
+	if w.firstWrite {
+		w.firstWrite = false
+		//Do not send noise for DNS requests just to be safe
+		if w.Writer.UDPOverride.Port != 53 {
+			var noise []byte
+			var err error
+			//User input string
+			if w.Writer.config.Noise.StrNoise != "" {
+				noise = []byte(w.Writer.config.Noise.StrNoise)
+			} else {
+				//Random noise
+				noise, err = GenerateRandomBytes(randBetween(int64(w.Writer.config.Noise.LengthMin),
+					int64(w.Writer.config.Noise.LengthMax)))
+			}
+
+			if err != nil {
+				return err
+			}
+			_, _ = w.Writer.PacketConnWrapper.Write(noise)
+
+			if w.Writer.config.Noise.DelayMin != 0 {
+				time.Sleep(time.Duration(randBetween(int64(w.Writer.config.Noise.DelayMin), int64(w.Writer.config.Noise.DelayMax))) * time.Millisecond)
+			}
+		}
+
+	}
+	for {
+		mb2, b := buf.SplitFirst(mb)
+		mb = mb2
+		if b == nil {
+			break
+		}
+		var n int
+		var err error
+		if b.UDP != nil {
+			if w.Writer.UDPOverride.Address != nil {
+				b.UDP.Address = w.Writer.UDPOverride.Address
+			}
+			if w.Writer.UDPOverride.Port != 0 {
+				b.UDP.Port = w.Writer.UDPOverride.Port
+			}
+			if w.Writer.Handler.config.hasStrategy() && b.UDP.Address.Family().IsDomain() {
+				ip := w.Writer.Handler.resolveIP(w.Writer.Context, b.UDP.Address.Domain(), nil)
+				if ip != nil {
+					b.UDP.Address = ip
+				}
+			}
+			destAddr, _ := net.ResolveUDPAddr("udp", b.UDP.NetAddr())
+			if destAddr == nil {
+				b.Release()
+				continue
+			}
+			n, err = w.Writer.PacketConnWrapper.WriteTo(b.Bytes(), destAddr)
```

Comment:
something is off, why does the noisewriter need to deal with ip resolution? it already wraps PacketWriter which has the same logic